controllers/users.js

```diff
@@ -611,11 +616,40 @@ const markUnverified = async (req, res) => {
       }
     });
 
+    const batchSize = 500;
+    const batches = Array.from({ length: Math.ceil(usersToApplyUnverifiedRole.length / batchSize) }, (_, index) =>
+      usersToApplyUnverifiedRole.slice(index * batchSize, index * batchSize + batchSize)
+    );
+
+    batches.forEach((batch) => {
+      const firestoreBatch = firestore.batch();
+
+      batch.forEach((id) => {
+        const memberRoleRef = memberRoleModel.doc(id);
+        const logRef = logsModel.doc();
+
+        firestoreBatch.set(memberRoleRef, {
+          roleid: unverifiedRoleId,
+          userid: id,
+          date: admin.firestore.Timestamp.fromDate(new Date()),
+        });
+
+        firestoreBatch.set(logRef, {
+          type: logType.ADD_UNVERIFIED_ROLE,
+          meta: { roleid: unverifiedRoleId, userid: id },
+          body: { message: "Unverified role added successfully" },
+          timestamp: admin.firestore.Timestamp.fromDate(new Date()),
+        });
+      });
+
+      batchPromises.push(firestoreBatch.commit());
+    });
+
     usersToApplyUnverifiedRole.forEach((id) => {
       addRolePromises.push(addRoleToUser(id, unverifiedRoleId));
     });
 
-    await Promise.all(addRolePromises);
+    await Promise.all([...addRolePromises, ...batchPromises]);
```

Comment:
Can it happen like some number of operation fails in a batch?

Reply:
If some operations within a batch insert to Firestore fail, the entire batch will fail, meaning none of the operations in the batch will be applied to the database